# Scripts and accessory data files for: A critical evaluation of *Mycobacterium bovis* pangenomics, with reference to its utility in outbreak investigation
Maintained by Kristina Ceres: kc649@cornell.edu

## Rscripts
* genome_quality_plot.R: quast and checkM results for assembly quality in **Supplemental Figure 1"
* Figure1_plots.R: scripts to create host/geographic distributions and map plot for **Figure 1**
* get_sample_lineages.R: assign M. bovis clonal complex labels from Zwyer et al. 
* PCA.R: Core and accessory genome clustering and plots for **Figure 2**
* ref_blast_matches.R: analyzing where M. bovis genes identifed by Panaroo map to the M. bovis reference af2122/97
* characterizing_acc_genes.R: plots for **Figure 3** related to differnet types of accessory genes
* essential_MTB_genes.R: determining if any pseudogenes were essential in Mtb
* process_Reis_and_Cunha_pangenome: looking at variable genes over the four outbreaks included in the study, and plot for **Figure 6**

## iTOL annotation files
* ITOL_glpK.txt: heatmap showing presence and absence of glpK 7C-->8C insertion used in **Supplemental Figure 4**
* ITOL_acc.txt: presence/absence of filtered accessory genes used in **Figure 4**
* redundantly_annotated_genes: directory for iTOL gradient files for **Supplementary Figure 3**

## tree files
* unfiltered_acc_gene_dendrogram.nwk: Jaccard distance dendrogram used in **Figure2A**
* core_genome.final_tree.tre: core phylogeny used in all figures that display core trees
* MN_SNPs.fasta.treefile: SNP tree used for **Figure 5**
* Glaser_MN_indels.fasta.treefile: indel tree for **Figure 5** 

## Parameter_file.md
* parameters for all bioinformatics not conducted in R




